pkg/kubelet/images/image_manager.go

```diff
@@ -224,3 +312,68 @@ func applyDefaultImageTag(image string) (string, error) {
 	}
 	return image, nil
 }
+
+// isEnsuredBySecret - returns two bools and error
+// 1. pulledBySecret: true if it is in ensured secret pulled image list,
+// the list will clean the image after PullImageSecretRecheckPeriod
+// 2. ensuredBySecret: true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image.
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (bool, bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if imageRef == "" {
+		return false, false, errors.New("imageRef is empty")
+	}
+
+	// if the image is in the ensured secret pulled image list, it is pulled by secret
+	pulledBySecret := m.ensureSecretPulledImages[imageRef] != nil
```

Comment:
checkpointing was out then in .. now out again for the alpha.. TODO: restore from checkpoint..

Reply:
> checkpointing was out then in .. now out again for the alpha.. TODO: restore from checkpoint..

Ok, sorry for asking questions which have already been answered :-)

Seems like checkpointing is the hard part of the problem. I'll defer to node approvers on that decision, but I'm a little surprised we're deferring tackling it.

Reply:
I was the one that proposed we drop the checkpointing. The motivation was the recheck period was proposed to be 24hr, and I think the kubelet restarting is an infrequent operation. Redundantly checking the creds is part of the feature's intention, so the admin is opting into checking more than necessary and that's the point. Not having checkpointing reduces the code complexity while not materially changing the behavior IMO

Reply:
If I'm reading correctly, the current implementation means images previously pulled to the node using pull secrets are treated as non-secret-pulled images after kubelet restart and would never be required to revalidate their pull credentials. Am I reading that correctly?

Reply:
I had a summary about the behavior in https://github.com/kubernetes/kubernetes/pull/114847#issuecomment-1857559176.
- Current Behavior: User has to use Always as current IfNotPresent will never do re-check.

Reply:
I think the problem @liggitt is describing is the ensureSecretPulledImages struct is only populated at the end of an EnsureImageExists call, which is after the first time it'd be used after the kubelet restarts. If that map is empty (because we have yet to try to pull an image) wouldn't the kubelet think it never needed to pull the image with a secret? an empty map is basically treated as no images required secret pulls, so the first time an image is requested after the recheck period or after kubelet restart would get no auth

Reply:
> wouldn't the kubelet think it never needed to pull the image with a secret?

if the gate is off .. ifNotPresent means, in the context of ensuring images for a new pod, don't recheck auth/manifests unless the tag is :latest  thus an admin must force pull always to check/recheck auth 

if the gate is on .. ifNotPresent means, in the context of ensuring images for a new pod, check/recheck within check option configuration parameters before starting a new pod as a part of the ensureImage process.. IOW a restart without checkpoint merely means we have to be more aggressive on the first pod ensureImage checks and go ahead and do the check to initialize the ensured/hash cache..

Reply:
Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all image IfNotPresent and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing. This case will need checkpointing to be implemented.

Reply:
> Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all images `IfNotPresent` and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing.

Exactly. This "how do we know which pre-existing images are intentionally preloaded and which are previously-pulled-by-secret" question is the hard part of the problem. That's why I was surprised to see this first version not handling it at all.

Reply:
> Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all images IfNotPresent and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing. This case will need checkpointing to be implemented.

this is true, but also if this is the case then that admin probably shouldn't enable this feature, because in addition to not handling kubelet restarts, there's also a recheck period. I suppose the recheck period could be arbitrarily high (never recheck) but it would be odd if a security minded admin wanted this feature but didn't utilize a lower frequency...

in other words: kubelet checkpointing could prevent some rechecks, but the point of this KEP is to add unnecessary rechecks. An admin should only enable it if their registries are expected to be available

Reply:
https://github.com/kubernetes/kubernetes/pull/94899/files#r691688013 prior discussion on this context

Reply:
> the point of this KEP is to add unnecessary rechecks

original point was to provide a path to get off pull always

The desire to do intermittent rechecks (e.g. after kubelet restart or after duration) for better security than if not present and better performance over pull always is at conflict with the desire to not have any rechecks because a recheck might fail in cases where a user does not use pull always or pull never and is expecting pull if not present to behave like pull never. 

With rechecking enabled for any reason in the scope of the if not present policy the checkpointing has minimal value.

With rechecking disabled for any reason in the scope of the if not present policy.. the checkpointing is needed before we go beta otherwise "how do we know which pre-existing images are intentionally preloaded and which are previously-pulled-by-secret."

So it would seem... the resolution for alpha would be to check if either of the recheck signals are enabled and if not .. disable the feature gate and possibly warn users that they should employ the pull always admission controller.

Reply:
https://github.com/kubernetes/kubernetes/pull/114847/commits/7a120ec8850900a419db335db37cf7063cba3b3f

if pullImageSecretRecheckPeriod>0 and KubeletEnsureSecretPulledImages is enabled, do recheck related logic.

This will keep the original behavior if we don't set the recheck period even when we enabled the FG by default in later releases or manually enable it in v1.30.

Does this make sense?

Reply:
I'm actually out of the office today, so I can't give this more time until next week, the more I look at this, I don't see how this can work as expected (reliably make new pods/pulls prove they have access AND not destabilize already running/working pods if a registry goes unavailable) without kubelet keeping track across restarts of images it used pull credentials to fetch.

There's two questions the kubelet has to answer:
1. did image X which is already pulled to the node require image pull credentials?
2. if so, are the image pull credentials this pod requesting image X is providing good?

This PR assumes all images that exist at kubelet start don't require credentials, only keeps track of credential-enabled image pulls the kubelet does in-memory, then forgets them on restart.

The "periodically check" approach looks like it will not actually protect images that exist on startup from use by uncredentialed pods. That seems like a fundamental gap.

The "periodically check" approach also looks like it will put existing working credentialed pods at risk of registry outage. Is it actually intended that pods which had image pull credentials and proved their pull credential was valid on their initial start on a node could stop working later because the registry is unavailable? That doesn't seem right.

https://github.com/kubernetes/kubernetes/pull/94899#discussion_r818144101 and https://github.com/kubernetes/kubernetes/pull/94899#issuecomment-1058126793 were the feedback on the last attempt of this feature, and looks like we're in essentially the same position now.

Reply:
The two stories from the kep https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#user-stories.
1. Story 1: User with multiple tenants will be able to support all image pull policies without concern that one tenant will gain access to an image that they don't have rights to.
2.Story 2: User will will no longer have to inject the Pull Always Image Pull Policy to ensure all tenants have rights to the images that are already present on a host.

The aim of the KEP is to gain an image pull policy that is between `Always` and current `IfNotPresent`.

- Image pull policy`Always` trust nothing.
- Image pull policy`IfNotPresent` trust everything on the disk, but no checking on the image pull secret.
- What we want is something like`IfNotPresentWithTrustedAuthPeriod`: 
  - we add a `recheck/trust period` to control the weak level: 24h is like a daily trust credit. 1y may be a long credit trust. Node admin can control its trust for an image pull credential check result.


The `PullImageSecretRecheckPeriod` naming is somewhat inaccurate, as we didn't recheck the image secret every 12h if it is set to be 12h. The check will happen only when a container startup or restart happen IIUC. It will not kill a pod or container if it found the credential is out-of-dated.

What should be done?
1. We should keep the behavior not changed even this feature is GAed.
2. When node admin options in with adding a `recheck period`, (in another word, this is for those admins who only allow its users using`Always` ImagePullPolicy now), I prefer that restarting kubelet needs recheck, which will be more secure for them. The admin can control the secure/weak level using the recheck period. This is also why I think no checkpoint is acceptable.